File: ACwing/codes/835.go
// go版本整个活，使用map去实现一个理论上支持所有字符的版本
package main

import "fmt"

type node struct {
	cnt  int
	next map[rune]*node
}

func newNode() *node {
	return &node{
		next: make(map[rune]*node),
	}
}

type trie = node

func NewTrie() *trie {
	return newNode()
}

func (t *trie) insert(s string) {
	rs := []rune(s)
	p := t
	for _, v := range rs {
		if _, ok := p.next[v]; !ok {
			p.next[v] = newNode()
		}
		p = p.next[v]
	}
	p.cnt++
}

func (t *trie) query(s string) int {
	p := t
	for _, r := range s {
		if _, ok := p.next[r]; !ok {
			return 0
		}
		p = p.next[r]
	}
	return p.cnt
}

func main() {
	n := 0
	fmt.Scanf("%d", &n)
	t := NewTrie()
	for i := 0; i < n; i++ {
		var op, s string
		fmt.Scanf("%s%s", &op, &s)
		if op == "I" {
			t.insert(s)
		} else {
			cnt := t.query(s)
			fmt.Println(cnt)
		}
	}
}

File: ACwing/codes/835.cpp
#include <iostream>
#include <string>
using namespace std;

constexpr int N = 1e5;
int son[N][26]; // 存储子节点指向哪里
int idx = 1; // 指向下一个被使用的缓冲区,注意要从1开始，0下标是根节点所在
int cnt[N]; // 存储以某个节点为结束的单词的出现的次数

void insert(const string& s) {
    int p = 0;
    for (const auto& c : s) {
        int a = c - 'a';
        if (son[p][a] == 0) { // 没有节点就创建节点
            son[p][a] = idx++;
        }
        p = son[p][a]; // 有节点就走向那个节点
    }
    cnt[p]++; // 标记节点结尾的单词的出现个数，也即某个串出现的个数。
}

int query(const string& s) {
    int p = 0;
    for (const auto& c : s) {
        int a = c - 'a';
        if (son[p][a] == 0) {
            return 0;
        }
        p = son[p][a];
    }
    return cnt[p];
}

int main() {
    int n;
    cin >> n;
    while (n--) {
        string op, s;
        cin >> op >> s;
        if (op == "I") {
            insert(s);
        } else {
            cout << query(s) << endl;
        }
    }
    return 0;
}
